### ISSUE TITLE
reverse mode

### ISSUE THREAD

**Issue by HajimeKawahara** [HajimeKawahara/exojax#67]
Currently exojax does not support custom VJP.

**Comment by HajimeKawahara** [comment]
related post: https://github.com/google/jax/issues/3996

**Comment by bmorris3** [comment]
I'm working on combining exojax with another model that might require reverse-mode diff, so I wonder if we can figure this out. Do you have any clues that might help me start to debug? As I'm sure you've experienced, the traceback when `forward_mode_differentiation=False` doesn't give many hints. 

What I'm seeing is that for reverse-mode diff, I'm getting a matrix reshape error that occurs when trying to turn a matrix of shape `[Nlayers x Nlines]` into one of shape `[Nlayers x Nlines x Nnu]`.

**Comment by bmorris3** [comment]
Ah I think I understand, so if you use [this function](https://github.com/HajimeKawahara/exojax/blob/67d1b6c868d69892d4bbf9e620ed05e432cfe61f/src/exojax/spec/rlpf.py#L153) `rxsmatrix`, instead of the standard `xsmatrix`, the reverse mode diff problem goes away?

**Comment by HajimeKawahara** [comment]
Yeah, that function is VJP custom-defined (not JPV). So, the reverse mode for [jax.experimental.optimizer](https://jax.readthedocs.io/en/latest/jax.experimental.optimizers.html), for instance, works. But, I think it does not work for NUTS in numpyro.

I do not have a useful hint, but [a simpler version of the reproduction](https://gist.github.com/HajimeKawahara/e490c44b66d6c719567835260500d86f) of the situation might be useful.

**Comment by bmorris3** [comment]
Does anything need to change for reverse-mode diff with continuum-opacity only calculations? I have a simple model that works in forward mode but doesn't seem to vary the input parameters in reverse-mode (but this may be my fault).

**Comment by bmorris3** [comment]
Update: it was my fault. Nevermind, sorry for the noise!

**Comment by HajimeKawahara** [comment]
[numpyro.plate](https://num.pyro.ai/en/stable/primitives.html#plate) should be used?

**Comment by dcmvdbekerom** [comment]
Hi, I was wondering if this issue is still open.. I tried your minimal example here https://gist.github.com/HajimeKawahara/e490c44b66d6c719567835260500d86f and both forward and reverse worked for me
jaxlib=0.1.75
jax=0.2.27
numpyro=0.8.0

**Comment by HajimeKawahara** [comment]
@dcmvdbekerom That's true! numpyro=0.7.2 also worked that example. So, now it's time to implement the reverse mode?

**Comment by HajimeKawahara** [comment]
According to Dirk's finding, I made a new branch jvp to try the reverse mode and found a new error w/ numpyro=0.8.0, but  may be a good sign:

``` sh
shirochan:~/exojax/tests/jvp(jvp)>python reverse_test.py 
xsmode assumes ESLOG: mode= lpf
WARNING: resolution may be too small. R= 382411.0658991112
Background atmosphere:  H2
Downloading http://www.exomol.com/db/CO/12C-16O/Li2015/12C-16O__Li2015.def
Downloading http://www.exomol.com/db/CO/12C-16O/Li2015/12C-16O__Li2015.pf
Downloading http://www.exomol.com/db/CO/12C-16O/Li2015/12C-16O__Li2015.states.bz2
Downloading http://www.exomol.com/db/CO/12C-16O/12C-16O__H2.broad
Downloading http://www.exomol.com/db/CO/12C-16O/12C-16O__He.broad
Downloading http://www.exomol.com/db/CO/12C-16O/12C-16O__air.broad
Error: Couldn't download .broad file and save.
Note: Couldn't find the hdf5 format. We convert data to the hdf5 format. After the second time, it will become much faster.
Reading transition file
Downloading http://www.exomol.com/db/CO/12C-16O/Li2015/12C-16O__Li2015.trans.bz2
Note: Couldn't find the hdf5 format. We convert data to the hdf5 format. After the second time, it will become much faster.
.broad is used.
Broadening code level= a0
H2-H2
Traceback (most recent call last):
  File "reverse_test.py", line 185, in <module>
    mcmc.run(rng_key_, nu1=nusd, y1=nflux)
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/numpyro-0.8.0-py3.8.egg/numpyro/infer/mcmc.py", line 599, in run
    states_flat, last_state = partial_map_fn(map_args)
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/numpyro-0.8.0-py3.8.egg/numpyro/infer/mcmc.py", line 387, in _single_chain_mcmc
    init_state = self.sampler.init(
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/numpyro-0.8.0-py3.8.egg/numpyro/infer/hmc.py", line 696, in init
    init_params = self._init_state(
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/numpyro-0.8.0-py3.8.egg/numpyro/infer/hmc.py", line 642, in _init_state
    init_params, potential_fn, postprocess_fn, model_trace = initialize_model(
  File "/home/kawahara/anaconda3/lib/python3.8/site-packages/numpyro-0.8.0-py3.8.egg/numpyro/infer/util.py", line 698, in initialize_model
    raise RuntimeError(
RuntimeError: Cannot find valid initial parameters. Please check your model again.
```

**Comment by dcmvdbekerom** [comment]
I get the same error

**Comment by HajimeKawahara** [comment]
I think this issue was solved by both update to numpyro 0.7 and #211.
